munit/shared/src/main/scala/munit/MUnitRunner.scala

```diff
@@ -279,61 +414,85 @@ class MUnitRunner(val cls: Class[_ <: Suite], newInstance: () => Suite)
       description: Description,
       test: suite.Test
   ): Future[Unit] = {
-    val result: Future[Any] = StackTraces.dropOutside {
+    val result = StackTraces.dropOutside {
       val beforeEachResult = runBeforeEach(test)
-      val any = beforeEachResult.error match {
-        case None =>
-          try test.body()
-          finally runAfterEach(test, beforeEachResult.loadedFixtures)
-        case Some(error) =>
-          try runAfterEach(test, beforeEachResult.loadedFixtures)
-          finally throw error
+      beforeEachResult.flatMap { r =>
+        r.error match {
+          case None =>
+            Future(test.body()).transformWithCompat {
+              case util.Failure(ex) =>
+                runAfterEach(test, r.loadedFixtures, r.loadedAsyncFixtures)
+                  .transformWithCompat(_ => Future.failed(ex))
+
+              case util.Success(testResult) =>
+                runAfterEach(test, r.loadedFixtures, r.loadedAsyncFixtures)
+                  .transformWithCompat {
+                    case util.Failure(ex) =>
+                      Future.failed(ex)
+
+                    case util.Success(_) =>
+                      Future.successful(testResult)
+                  }
+            }
+
+          case Some(ex) =>
+            runAfterEach(test, r.loadedFixtures, r.loadedAsyncFixtures)
+              .transformWithCompat(_ => Future.failed(ex))
+        }
       }
-      futureFromAny(any)
     }
-    result.map {
-      case f: TestValues.FlakyFailure =>
+
+    result
+      .map {
+        case TestValues.Ignore =>
+          notifier.fireTestIgnored(description)
+        case _ =>
+          ()
+      }
+      .recover { case f: TestValues.FlakyFailure =>
         trimStackTrace(f)
         notifier.fireTestAssumptionFailed(new Failure(description, f))
-      case TestValues.Ignore =>
-        notifier.fireTestIgnored(description)
-      case _ =>
-        ()
-    }
+      }
   }
 
-  private def foreachUnsafe(thunks: Iterable[() => Unit]): Try[Unit] = {
-    var errors = mutable.ListBuffer.empty[Throwable]
-    thunks.foreach { thunk =>
-      try {
-        thunk()
-      } catch {
-        case ex if NonFatal(ex) =>
-          errors += ex
-      }
-    }
-    errors.toList match {
-      case head :: tail =>
-        tail.foreach { e =>
-          if (e ne head) {
-            head.addSuppressed(e)
+  private def foreachUnsafe[A](
+      thunks: Seq[(A, () => Future[Unit])]
+  ): Future[Either[(Throwable, List[A]), List[A]]] =
+    for {
+      computed <- Future.sequence(
+        thunks.map { case (fixture, computation) =>
+          computation().map(_ => Right[Throwable, A](fixture)).recover {
+            case ex if NonFatal(ex) =>
+              Left[Throwable, A](ex)
           }
         }
-        scala.util.Failure(head)
-      case _ =>
-        scala.util.Success(())
-    }
-  }
+      )
+
+      aggregated = computed.foldLeft(List.empty[Throwable] -> List.empty[A]) {
```

Comment:
We should avoid methods like `foldLeft` since they increase the stack trace. It's better to write out the recursion by hand with `@tailrec`

Reply:
My bad, `List.foldLeft` doesn't have this problem. I still prefer manually written tailrec methods over `foldLeft`

Reply:
I'm in love with `fold`s. But for keeping consistency in the codebase have rewritten it to the tail-recursion function.